File: 081.Search in Rotated Sorted Array II.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        return self.find(nums, target, 0, len(nums) - 1)
        
    def find(self, nums, target, start, end):
        if start > end:
            return False
        mid = (start + end) / 2
        if nums[mid] == target:
            return True
        if nums[start] > nums[end] or nums[start] == nums[end]:
            ret = self.find(nums, target, start, mid - 1)
            if not ret:
                ret = self.find(nums, target, mid + 1, end)
            return ret
        if nums[mid] > target:
            return self.find(nums, target, start, mid - 1)
        else:
            return self.find(nums, target, mid + 1, end)

File: 081.Search in Rotated Sorted Array II.java
/*
    Follow up for "Search in Rotated Sorted Array":
    What if duplicates are allowed?

    Would this affect the run-time complexity? How and why?
    Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

    (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

    Write a function to determine if a given target is in the array.

    The array may contain duplicates.
     */

    public boolean search(int[] nums, int target) {
        int low = 0, high = nums.length-1;
        if (high < 0)
            return false;
        int middle;
        while (low < high) {
            middle = (low + high) / 2;
            if (nums[middle] == target)
                return true;
            if (nums[low] < nums[middle]) {
                if (nums[middle] > target && nums[low] <= target)
                    high = middle;
                else
                    low = middle + 1;
            } else if (nums[middle] < nums[high]) {
                if (nums[middle] < target && nums[high] >= target)
                    low = middle + 1;
                else
                    high = middle;
            } else {
                if (nums[high] == target)
                    return true;
                else
                    high--;
            }
        }

        return nums[low] == target;
    }
